docs/ko/technical/webhook-server-internal.md

```diff
@@ -0,0 +1,66 @@
+이 문서는 웹후크 기능의 내부적 동작에 대해 설명한다.
+
+개요
+--------
+
+Yobi의 웹후크 기능은 프로젝트 내의 특정 이벤트(푸시, 코드 주고받기, 이슈 생성 등)와 이로 인해 발생하는 HTTP 요청을 통해 이루어진다. 현재는 push 이벤트에 대해 웹후크가 구현되어 있다.
+
+웹후크의 등록과 삭제는 HTTP 요청과 응답을 통해 이루어진다. 요청은 `controller.ProjectApp`이 `models.Webhook`을 이용하여 처리하며, 해당 요청은 conf/routes 파일에 다음과 같이 정의되어 있다.
+
+```
+    GET     /:user/:project/webhooks              controllers.ProjectApp.webhooks(user, project)
+    POST    /:user/:project/webhooks              controllers.ProjectApp.newWebhook(user, project)
+    POST    /:user/:project/webhooks/:id          controllers.ProjectApp.deleteWebhook(user, project, id: Long)
+```
+
+웹후크 등록 및 삭제
+--------
+
+웹후크의 등록 및 등록한 웹후크의 확인은 `/:user/:project/webhooks`으로 route되는 view를 통해 할 수 있다. 등록된 웹후크는 해당 웹후크가 등록된 프로젝트 내에서만 확인하고 관리할 수 있으며, 해당 프로젝트에서 발생하는 이벤트를 통해서만 작동한다.
+
+`/:user/:project/webhooks`으로 route되는 view는 등록된 웹후크의 리스트를 보여주며, 웹후크의 등록과 삭제를 POST 요청을 통해 처리하는 UI를 제공한다.
+
+새로운 웹후크의 등록은 `/:user/:project/webhooks`으로 HTTP POST 요청을 함으로서 이루어진다. 이 때 폼을 통해 두 종류의 입력을 payload로 받게 된다. 해당 payload의 명세는 다음과 같다.
+* Payload URL : 웹후크가 동작할 시 HTTP 요청을 보낼 주소를 입력한다. URL의 길이는 2000바이트로 제한한다.
+* Secret : HTTP 요청을 받을 서버에서 해당 요청을 구분하기 위한 비밀 토큰을 입력한다. 토큰의 길이는 250바이트로 제한한다.
+
+웹후크가 정상적으로 등록되면, 서버는 view를 새로고침하여 추가된 웹후크가 view에 정상적으로 반영되도록 한다. 웹후크 등록에 대한 실제 코드는 `model.Webhook.create`에서 확인할 수 있다.
+
+등록된 웹후크의 삭제는 `/:user/:project/webhooks/:id` 로 HTTP POST 요청을 함으로서 이루어지며, id가 :id인 웹후크를 삭제한다. 웹후크 삭제에 대한 실제 코드는 `model.Webhook.delete`에서 확인할 수 있다.
+
+
+웹후크 작동
+--------
+
+등록된 웹후크는 웹후크가 등록된 프로젝트에서 특정 이벤트가 발생했을때 작동한다.
+
+프로젝트 내에서 발생하는 이벤트에 따른 웹후크의 작동 명세는 `app.playRepository.hooks.RequestByWebhooks`에 정의되어 있다. 현재는 해당 hook이 `app.playRepository.RepositoryService.PostReceiveHook`에 등록이 되어있어 push 이벤트에 대해 반응을 하며, 향후 이벤트의 종류를 확장할 시 이와 같이 `RequestByWebhooks`을 필요한 이벤트에 걸어줌으로서 알맞게 대응할 수 있다.
+
+웹후크가 작동할 시, 서버는 웹후크에 등록된 Payload URL로 HTTP POST 요청을 전송한다. 요청을 보낼 시 payload에는 웹후크 이벤트의 정보가 담긴 JSON object가 포함된다. 전송되는 JSON object의 예시는 아래와 같다.
+```
+{
+    "hook_id":41,
+    "hook":{
+        "id":41,
+        "name":"web",
+        "active":true,
+        "events":[
+            "push"
+        ],
+        "config":{
+            "url":"http://your.url/for/webhook",
+            "content_type":"json",
+            "secret":"your-webhook-token"
+        },
+        "created_at":"2015-06-05T12:42:43Z"
+    },
+    "repository":{
+        "id":33,
+        "name":"your-project",
+        "owner":"your-user-name",
+        "html_url":"http://url-of.yobi:9000/project-name",
+        "overview":"project-overview"
+    }
+}
```

Comment:
착각이 아니군요. 실제로 github이 보내는 payload하고 다르네요.

github의 예:
```
POST / HTTP/1.1
Host: npcode.com:1234
Accept: */*
User-Agent: GitHub-Hookshot/99ae70d
X-GitHub-Event: push
X-GitHub-Delivery: 45582480-0fe5-11e5-9d5d-e647bbf8699a
content-type: application/json
X-Hub-Signature: sha1=b2d755ffe246a7e762a222002e7bb03280eb8d9d
Content-Length: 5434

{"ref":"refs/heads/master","before":"229d605a20ba198bb925de6bbcc0a5f4805709a2","after":"1285ea25b4b41f3a4c61ee425f4135ca2e9a0d17","created":false,"deleted":false,"forced":false,"base_ref":null,"compare":"https://github.com/npcode/foo/compare/229d605a20ba...1285ea25b4b4","commits":[{"id":"1285ea25b4b41f3a4c61ee425f4135ca2e9a0d17","distinct":true,"message":"gggg","timestamp":"2015-06-11T11:55:01+09:00","url":"https://github.com/npcode/foo/commit/1285ea25b4b41f3a4c61ee425f4135ca2e9a0d17","author":{"name":"Yi EungJun","email":"eungjun.yi@navercorp.com","username":"npcode"},"committer":{"name":"Yi EungJun","email":"eungjun.yi@navercorp.com","username":"npcode"},"added":[],"removed":[],"modified":["a"]}],"head_commit":{"id":"1285ea25b4b41f3a4c61ee425f4135ca2e9a0d17","distinct":true,"message":"gggg","timestamp":"2015-06-11T11:55:01+09:00","url":"https://github.com/npcode/foo/commit/1285ea25b4b41f3a4c61ee425f4135ca2e9a0d17","author":{"name":"Yi EungJun","email":"eungjun.yi@navercorp.com","username":"npcode"},"committer":{"name":"Yi EungJun","email":"eungjun.yi@navercorp.com","username":"npcode"},"added":[],"removed":[],"modified":["a"]},"repository":{"id":15792441,"name":"foo","full_name":"npcode/foo","owner":{"name":"npcode","email":"semtlenori@gmail.com"},"private":false,"html_url":"https://github.com/npcode/foo","description":"test","fork":false,"url":"https://github.com/npcode/foo","forks_url":"https://api.github.com/repos/npcode/foo/forks","keys_url":"https://api.github.com/repos/npcode/foo/keys{/key_id}","collaborators_url":"https://api.github.com/repos/npcode/foo/collaborators{/collaborator}","teams_url":"https://api.github.com/repos/npcode/foo/teams","hooks_url":"https://api.github.com/repos/npcode/foo/hooks","issue_events_url":"https://api.github.com/repos/npcode/foo/issues/events{/number}","events_url":"https://api.github.com/repos/npcode/foo/events","assignees_url":"https://api.github.com/repos/npcode/foo/assignees{/user}","branches_url":"https://api.github.com/repos/npcode/foo/branches{/branch}","tags_url":"https://api.github.com/repos/npcode/foo/tags","blobs_url":"https://api.github.com/repos/npcode/foo/git/blobs{/sha}","git_tags_url":"https://api.github.com/repos/npcode/foo/git/tags{/sha}","git_refs_url":"https://api.github.com/repos/npcode/foo/git/refs{/sha}","trees_url":"https://api.github.com/repos/npcode/foo/git/trees{/sha}","statuses_url":"https://api.github.com/repos/npcode/foo/statuses/{sha}","languages_url":"https://api.github.com/repos/npcode/foo/languages","stargazers_url":"https://api.github.com/repos/npcode/foo/stargazers","contributors_url":"https://api.github.com/repos/npcode/foo/contributors","subscribers_url":"https://api.github.com/repos/npcode/foo/subscribers","subscription_url":"https://api.github.com/repos/npcode/foo/subscription","commits_url":"https://api.github.com/repos/npcode/foo/commits{/sha}","git_commits_url":"https://api.github.com/repos/npcode/foo/git/commits{/sha}","comments_url":"https://api.github.com/repos/npcode/foo/comments{/number}","issue_comment_url":"https://api.github.com/repos/npcode/foo/issues/comments{/number}","contents_url":"https://api.github.com/repos/npcode/foo/contents/{+path}","compare_url":"https://api.github.com/repos/npcode/foo/compare/{base}...{head}","merges_url":"https://api.github.com/repos/npcode/foo/merges","archive_url":"https://api.github.com/repos/npcode/foo/{archive_format}{/ref}","downloads_url":"https://api.github.com/repos/npcode/foo/downloads","issues_url":"https://api.github.com/repos/npcode/foo/issues{/number}","pulls_url":"https://api.github.com/repos/npcode/foo/pulls{/number}","milestones_url":"https://api.github.com/repos/npcode/foo/milestones{/number}","notifications_url":"https://api.github.com/repos/npcode/foo/notifications{?since,all,participating}","labels_url":"https://api.github.com/repos/npcode/foo/labels{/name}","releases_url":"https://api.github.com/repos/npcode/foo/releases{/id}","created_at":1389342319,"updated_at":"2014-06-10T07:17:27Z","pushed_at":1433991308,"git_url":"git://github.com/npcode/foo.git","ssh_url":"git@github.com:npcode/foo.git","clone_url":"https://github.com/npcode/foo.git","svn_url":"https://github.com/npcode/foo","homepage":null,"size":144,"stargazers_count":0,"watchers_count":0,"language":null,"has_issues":true,"has_downloads":true,"has_wiki":true,"has_pages":false,"forks_count":0,"mirror_url":null,"open_issues_count":1,"forks":0,"open_issues":1,"watchers":0,"default_branch":"master","stargazers":0,"master_branch":"master"},"pusher":{"name":"npcode","email":"semtlenori@gmail.com"},"sender":{"login":"npcode","id":1129852,"avatar_url":"https://avatars.githubusercontent.com/u/1129852?v=3","gravatar_id":"","url":"https://api.github.com/users/npcode","html_url":"https://github.com/npcode","followers_url":"https://api.github.com/users/npcode/followers","following_url":"https://api.github.com/users/npcode/following{/other_user}","gists_url":"https://api.github.com/users/npcode/gists{/gist_id}","starred_url":"https://api.github.com/users/npcode/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/npcode/subscriptions","organizations_url":"https://api.github.com/users/npcode/orgs","repos_url":"https://api.github.com/users/npcode/repos","events_url":"https://api.github.com/users/npcode/events{/privacy}","received_events_url":"https://api.github.com/users/npcode/received_events","type":"User","site_admin":false}}
```